contract/contracts/RockPaperScissors.aes

```diff
@@ -0,0 +1,167 @@
+    Chain.event(Player0ProvidedHash(hash,Call.value))
+
+  payable stateful entrypoint player1_move(move_str: string): unit =
+    require_player1()
+    ensure_player1_turn_to_move()
+    let move = str_to_move(move_str)
+    require(Call.value == state.stake,
+      String.concat("wrong_stake, expected ", Int.to_str(state.stake)))
+    put(state{ player1_move  = Some(move),
+               last_move_timestamp        = get_timestamp()})
+    Chain.event(Player1Moved(move_str,Call.value))
+
+  stateful entrypoint reveal(key: string, move_str: string): option(address) =
+    require_player0()
+    ensure_player0_turn_to_reveal()
+    let player0_move = str_to_move(move_str)
+    ensure_if_key_is_valid(key, move_str)
+    let Some(player1_move) = state.player1_move
+    //TODO:  get the winner
+    let stake = state.stake
+    Chain.event(Player0Revealed(move_str))
+    reset_state()
+    let (p0,p1) = (state.player0,state.player1)
+    switch(get_winner(player0_move,player1_move))
+        None =>
+          let p0_refund = Contract.balance / 2
+          Chain.spend( p0, p0_refund )
+          let p1_refund = Contract.balance
+          Chain.spend( p1, p1_refund) // send the rest of it
+          Chain.event(Draw(p0_refund,p1_refund,
+              String.concat(
+                Address.to_str(p0),
+                String.concat("|",Address.to_str(p1))
+                )
+            ))
+          None
+        Some(true) =>
+          Chain.event(Player0Won(p0,Contract.balance))
+          Chain.spend(p0, Contract.balance)
+          Some(p0)
+        _ =>
+          Chain.event(Player1Won(p1,Contract.balance))
+          Chain.spend(p1, Contract.balance)
+          Some(p1)
+
+  stateful entrypoint player1_dispute_no_reveal(): unit =
+    require_player1()
+    ensure_player0_turn_to_reveal()
+    require(state.last_move_timestamp  + state.reaction_time < get_timestamp(), "not_yet_allowed")
+    Chain.event(Player1WonDispute(state.player1,Contract.balance))
+    Chain.spend(state.player1, Contract.balance)
+    reset_state()
+
+  stateful entrypoint player0_dispute_no_move(): unit =
+    require_player0()
+    ensure_player1_turn_to_move()
+    require(state.last_move_timestamp + state.reaction_time < get_timestamp(), "not_yet_allowed")
+    Chain.event(Player0WonDispute(state.player0,Contract.balance))
+    Chain.spend(state.player0, Contract.balance)
+    reset_state()
+
+  entrypoint compute_hash(key: string, move: string) : hash =
+    String.sha256(String.concat(key, move))
+
+  //for debug / test purposes
+  stateful entrypoint set_timestamp(timestamp: int ): unit =
+    require(state.debug_timestamp != None,"not_debug_mode")
+    require(timestamp >= 0,"not_positive")
+    put(state { debug_timestamp = Some(timestamp) })
+
+  // internal functions
+  function get_timestamp() = Option.default(Chain.timestamp,state.debug_timestamp)
+
+  function ensure_player1_turn_to_move() =
+    require(state.hash != None, "no_hash")
```

Comment:
isn't this incorrect? I assume player0 is the state channel computer, doesn't the hash has to be provided first, otherwise player0 would know the move of player1 before posting the player0 move?

Reply:
Yes, the hash is provided  first by the payer0, which means until that moment the `hash == None`. So the player1 can't move, he has to wait the player0 to add the hash (`state.hash != None`). This function succeeds if it is the player1's turn to move

Reply:
Whoops my bad!